### PR TITLE
Conditional Guard Description

### PR DESCRIPTION
**Pull request by jgorak**
### Description

Custom description option (:desc) on only_if/not_if guards for better identification in output
```
log "hello" do
  not_if(:desc => 'randomly skip') { rand < 0.5 }
end
```

### Issues Resolved

https://github.com/chef/chef/issues/6322

### Check List

- [x] New functionality includes tests
- [x] All tests pass
- [x] RELEASE\_NOTES.md, has been updated if required (not required for bugfixes, required for API changes)
- [x] All commits have been signed-off for the Developer Certificate of Origin. See <https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco>

### CODE REVIEW

**Comment by coderanger** [comment]
Something this big I would want to see go through the RFC process. Personally I think this is overkill, code this complex can use `shell_out` manually to do the same thing.

**Comment by jgorak** [comment]
@coderanger Thanks for the feedback.  I'm not sure I understand the shell_out reference, though.  Would you be able to shoe me how it could be used to replace the example at the top?

File: lib/chef/resource.rb [hunk 1]
```diff
@@ -416,6 +420,10 @@ def only_if(command = nil, opts = {}, &block)
     # @param block [Proc] A ruby block to run. Ignored if a command is given.
     #
     def not_if(command = nil, opts = {}, &block)
+      if command.is_a?(Hash) && opts.empty?
+        opts = command
+        command = nil
+      end
```
Comment (lamont-granquist) [inline lib/chef/resource.rb]: does `def not_if(command = nil, **opts, &block)` work here without this mangling, because this code seems brittle to me...

**Comment by coderanger** [comment]
@jgorak You can do something like this:

```ruby
log "hello" do
  not_if do
    (rand < 0.5).tap do |val|
      Chef::Log.info("a thing!") unless val
    end
end
```

And similarly in the case of `shell_out(...).error?` as the condition for a command string.

File: lib/chef/resource.rb (hunk 1 above)
Comment (jgorak) [inline lib/chef/resource.rb]: I agree 100%.  I use ruby 1.9 on my daily, and really just never realized double-splat worked a little differently than the way k:v pairs implicitly load into opts={} .

Thanks for the feedback.

**Comment by jgorak** [comment]
@coderanger I see.  Thought I was missing some trick that might make it cleaner.  That works, though in my opinion it's not as simple/clean as it could potentially be.

Thanks again

**Comment by coderanger** [comment]
@jgorak I agree it's not as nice, but this is clearly not something that has a huge need or others would have wanted it by now so for a minor hit to complexity of the recipe code, we can avoid adding the complexity to Chef core. The main case I would imagine is identifying between multiple guard clauses on a single resource, which is itself something we caution against as it leads to code that is very difficult to read.

File: lib/chef/resource.rb (hunk 1 above)
Comment (lamont-granquist) [inline lib/chef/resource.rb]: yeah that was going to be my other comment is that i would have expected to see something like `opts = opts.dup.delete(:desc)` here.

**Comment by lamont-granquist** [comment]
closing due to inactivity

**Pull request closed without merging**